c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1546,7 +1551,8 @@ class DeviceCachingAllocator {
       // splitting a block depends on `max_split_size`, which may have changed
       // between whe checkpoint was taken and now, so we make sure to recreate
       // the behavior from the checkpoint.
-      bool split = (i + 1) < segment.blocks.size();
+      bool split = (i + 1) < segment.blocks.size() &&
```

Comment:
The comment above is still relevant: allocator settings may be different at the time should_split is called here and may not restore the split in the way that it was when the checkpoint was taken.

Reply:
makes sense, replaced w a check to prevent splitting a block to create a new block of size 0, which was happening w expandable segments turned on

Reply:
I think the condition `(i + 1) < segment.blocks.size()` is now redundant. We can tell if we need to split block by asking if there is any remaining space in the segment after putting in the blocks we already have.